config.go

```diff
@@ -124,27 +85,53 @@ type SpanOption interface {
 	ApplySpan(*SpanConfig)
 }
 
+// NewEventConfig applies all the EventOptions to a returned SpanConfig. If no
+// timestamp option is passed, the returned SpanConfig will have a Timestamp
+// set to the call time, otherwise no validation is performed on the returned
+// SpanConfig.
+func NewEventConfig(options ...EventOption) *SpanConfig {
+	c := new(SpanConfig)
+	for _, option := range options {
+		option.ApplyEvent(c)
+	}
+	if c.Timestamp.IsZero() {
+		c.Timestamp = time.Now()
+	}
+	return c
+}
+
+// EventOption applies span life-cycle event options to a SpanConfig.
+type EventOption interface {
+	SpanOption
+
+	ApplyEvent(*SpanConfig)
+}
+
 type attributeSpanOption []label.KeyValue
 
-func (o attributeSpanOption) ApplySpan(c *SpanConfig) {
+func (o attributeSpanOption) ApplySpan(c *SpanConfig)  { o.apply(c) }
+func (o attributeSpanOption) ApplyEvent(c *SpanConfig) { o.apply(c) }
+func (o attributeSpanOption) apply(c *SpanConfig) {
 	c.Attributes = append(c.Attributes, []label.KeyValue(o)...)
 }
 
 // WithAttributes adds the attributes to a span. These attributes are meant to
 // provide additional information about the work the Span represents. The
 // attributes are added to the existing Span attributes, i.e. this does not
 // overwrite.
-func WithAttributes(attributes ...label.KeyValue) SpanOption {
+func WithAttributes(attributes ...label.KeyValue) EventOption {
 	return attributeSpanOption(attributes)
 }
 
 type timestampSpanOption time.Time
 
-func (o timestampSpanOption) ApplySpan(c *SpanConfig) { c.Timestamp = time.Time(o) }
+func (o timestampSpanOption) ApplySpan(c *SpanConfig)  { o.apply(c) }
+func (o timestampSpanOption) ApplyEvent(c *SpanConfig) { o.apply(c) }
+func (o timestampSpanOption) apply(c *SpanConfig)      { c.Timestamp = time.Time(o) }
 
 // WithTimestamp sets the time of a Span life-cycle moment (e.g. started or
 // stopped).
-func WithTimestamp(t time.Time) SpanOption {
+func WithTimestamp(t time.Time) EventOption {
```

Comment:
```suggestion
func WithTimestamp(t time.Time) timestampSpanOption {
```

Should this return the concrete type rather than a specific interface?

Reply:
That's a good question, but overall I think this should still return the related interface (though maybe a different one). Doing so follows our standard options pattern and as a nicity ensures at compile time that the struct implements the EventOption interface.

As for documentation it will group this with the `EventOption` instead of the `SpanOption`:

![20201013_160648](https://user-images.githubusercontent.com/5543599/95925238-23049a80-0d6e-11eb-9073-003fc4605333.png)

However if we return the unexported type it has no association:

![20201013_160825](https://user-images.githubusercontent.com/5543599/95925317-56dfc000-0d6e-11eb-81a4-2b6206164252.png)

This alternative gives less information to the users as there is no documentation for `timestampSpanOption` and they will need to look into the implementation.  Also, by associating with the `EventOption` the documentation for that interface suggests through its composition that it can be used for all `SpanOption` interfaces.

That said I took a shortcut here and made the `EventOption` an "extension" of the `SpanOption`. There is also the option to have an `EventOption` interface that only works for events and then have a "combined" interface [like we normally do](https://github.com/open-telemetry/opentelemetry-go/blob/master/CONTRIBUTING.md#dealing-with-overlap). E.g.

```
type LifeCycleOption interface {
	SpanOption
	EventOption
}
```

And these functions can return the struct as that interface type.

The downside is that there will (currently) be no direct `EventOptions`. Though, I guess that could definitely change in the future :shrug:.

Reply:
Bleh! I'm not sure I like any of these options.  Having `EventOption` extend `SpanOption` and grouping the method there may leave people confused as to whether it  can be used as a `SpanOption` unless they pay close attention to the definition of `EventOption`.  I wonder whether it might be better to return the concrete type from the `With*()` functions and have a consistent documentation format for indicating which `*Option` interfaces they satisfy.  That should result in all of the `With*()` functions grouped together alphabetically in the package-level function section with each describing its scope of applicability.

Reply:
From the SIG meeting: switch to the composite `LifeCycleOption` here and open an issue to look into returning the unexported type from option creation functions as a part of our standards.

Reply:
Updated PTAL